server/src/main/java/org/elasticsearch/action/update/TransportUpdateAction.java

```diff
@@ -333,4 +343,58 @@ private void handleUpdateFailureWithRetry(
         }
         listener.onFailure(cause instanceof Exception ? (Exception) cause : new NotSerializableExceptionWrapper(cause));
     }
+
+    private static UpdateHelper.Result deleteInferenceResults(
+        UpdateRequest updateRequest,
+        UpdateHelper.Result result,
+        IndexMetadata indexMetadata,
+        MappingLookup mappingLookup
+    ) {
+        if (result.getResponseResult() != DocWriteResponse.Result.UPDATED) {
+            return result;
+        }
+
+        Map<String, InferenceFieldMetadata> inferenceFields = indexMetadata.getInferenceFields();
+        if (inferenceFields.isEmpty()) {
+            return result;
+        }
+
+        Map<String, Object> updateRequestSource = updateRequest.doc().sourceAsMap();
+        Map<String, Object> updatedSource = result.updatedSourceAsMap();
+        boolean updatedSourceModified = false;
+        for (var entry : inferenceFields.entrySet()) {
+            String inferenceFieldName = entry.getKey();
+            Mapper mapper = mappingLookup.getMapper(inferenceFieldName);
+
+            if (mapper instanceof InferenceFieldMapper inferenceFieldMapper) {
+                String[] sourceFields = entry.getValue().getSourceFields();
+                for (String sourceField : sourceFields) {
+                    if (sourceField.equals(inferenceFieldName) == false && updateRequestSource.containsKey(sourceField)) {
+                        // Replace the inference field's value with its original value (i.e. the user-specified value).
+                        // This has two important side effects:
+                        // - The inference field value will remain parsable by its mapper
+                        // - The inference results will be removed, forcing them to be re-generated downstream
+                        updatedSource.put(inferenceFieldName, inferenceFieldMapper.getOriginalValue(updatedSource));
```

Comment:
We replace the inference field's value with its original value here to force inference results to be re-generated downstream

Reply:
In case the original value is null we should remove the field, otherwise the bulk action should fail no?

Reply:
ah ok got it, so we do keep the inference results if none of the source field are present in the update, nice!

Reply:
I _think_ we're ok with a `null` field value because updates through the Update API generate an index bulk request, and those don't require a value for every source field. See this YAML test: https://github.com/Mikep86/elasticsearch/blob/3f687a5997289950f8feef9845c5846c531dac4f/x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/30_semantic_text_inference.yml#L491